File: FutureVoice/Services/DrillReminder.swift
import Foundation
import UserNotifications

/// One local notification for the drill queue — SRS only works if something
/// pulls the user back when cards come due, and nothing did until now.
///
/// Policy (deliberately quiet):
///   • At most ONE pending reminder at any time, always rescheduled from the
///     current queue state — never a stack of stale notifications.
///   • Fires when the next card becomes due, clamped into 09:00–21:00 local
///     time so a 3-day Leitner interval landing at 2am waits for morning.
///   • If cards are ALREADY due while the user is in the app, remind
///     tomorrow morning instead — nudging about a queue they can see is noise.
///   • Permission is requested only from a foreground moment right after new
///     cards were created (post-session), never from the background path.
@MainActor
enum DrillReminder {

    private static let requestId = "futurevoice.drill-due"
    private static let dayStartHour = 9
    private static let dayEndHour = 21

    /// Recompute and replace the pending reminder from current queue state.
    /// - Parameter allowPermissionPrompt: pass true only from a foreground,
    ///   contextual moment (e.g. right after a session created cards).
    static func reschedule(allowPermissionPrompt: Bool = false, now: Date = Date()) async {
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: [requestId])

        let cards = DrillStore.shared.load()
        guard !cards.isEmpty else { return }

        switch await center.notificationSettings().authorizationStatus {
        case .notDetermined:
            guard allowPermissionPrompt else { return }
            let granted = (try? await center.requestAuthorization(options: [.alert, .sound])) ?? false
            guard granted else { return }
        case .denied:
            return
        default:
            break
        }

        let hasDueNow = cards.contains { $0.nextReviewAt <= now }
        let nextFutureDue = cards.map(\.nextReviewAt).filter { $0 > now }.min()
        guard let fireDate = fireDate(now: now, nextDue: nextFutureDue, hasDueNow: hasDueNow),
              fireDate > now else { return }

        // How many cards will be waiting at fire time.
        let countAtFire = cards.filter { $0.nextReviewAt <= fireDate }.count
        guard countAtFire > 0 else { return }

        let content = UNMutableNotificationContent()
        content.title = "Drills are due"
        content.body = countAtFire == 1
            ? "1 phrase is ready to review."
            : "\(countAtFire) phrases are ready to review."
        content.sound = .default

        let comps = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute], from: fireDate)
        let trigger = UNCalendarNotificationTrigger(dateMatching: comps, repeats: false)
        let request = UNNotificationRequest(identifier: requestId, content: content, trigger: trigger)
        try? await center.add(request)
    }

    /// Picks when to fire, clamped into the waking-hours window.
    static func fireDate(
        now: Date,
        nextDue: Date?,
        hasDueNow: Bool,
        calendar: Calendar = .current
    ) -> Date? {
        let candidate: Date
        if hasDueNow {
            // User can see today's queue right now — remind tomorrow morning.
            guard let tomorrow = calendar.date(byAdding: .day, value: 1, to: now),
                  let morning = calendar.date(bySettingHour: dayStartHour, minute: 0, second: 0, of: tomorrow)
            else { return nil }
            candidate = morning
        } else if let nextDue {
            candidate = nextDue
        } else {
            return nil
        }

        // A due time inside the next 12 hours can only have come from the
        // learner picking one in the bin tray — every ladder interval is a
        // day or longer, and box 0 is "due now" (handled above). They asked
        // for it, at an hour they were awake to ask: fire exactly then rather
        // than parking a 10-minute snooze until 9am.
        if candidate.timeIntervalSince(now) < 12 * 60 * 60 { return candidate }

        let hour = calendar.component(.hour, from: candidate)
        if hour < dayStartHour {
            return calendar.date(bySettingHour: dayStartHour, minute: 0, second: 0, of: candidate)
        }
        if hour >= dayEndHour {
            guard let nextDay = calendar.date(byAdding: .day, value: 1, to: candidate) else { return nil }
            return calendar.date(bySettingHour: dayStartHour, minute: 0, second: 0, of: nextDay)
        }
        return candidate
    }
}

File: android/app/src/main/java/com/roro/futurevoice/data/DrillReminder.kt
package com.roro.futurevoice.data

import android.app.AlarmManager
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.app.NotificationCompat
import com.roro.futurevoice.MainActivity
import com.roro.futurevoice.R
import java.util.Calendar

/**
 * One local notification for the review queue — spaced repetition only works
 * if something pulls the learner back when cards come due, and the per-item
 * callbacks ([ReviewQueue]) answer "you asked to see THIS again", not
 * "something is waiting".
 *
 * Policy (deliberately quiet, iOS `DrillReminder`):
 *   • At most ONE pending reminder, always rebuilt from the current queue —
 *     never a stack of stale notifications.
 *   • Fires when the next card becomes due, clamped into 09:00–21:00 local
 *     time so a 3-day interval landing at 2am waits for morning.
 *   • If cards are ALREADY due while the learner is in the app, it waits for
 *     tomorrow morning — nudging about a queue they can see is noise.
 */
object DrillReminder {
    private const val REQUEST_CODE = 0x44_000001
    private const val NOTIFICATION_ID = 0x44_000002
    private const val DAY_START_HOUR = 9
    private const val DAY_END_HOUR = 21

    /** Recompute and replace the pending reminder from the current queue —
     *  drill cards AND the words/expressions the daily deck snoozed. */
    suspend fun reschedule(context: Context, now: Long = System.currentTimeMillis()) {
        val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        am.cancel(pending(context, 0))
        val cards = runCatching { DrillStore.shared(context).load().map { it.nextReviewAt } }.getOrDefault(emptyList())
        val study = runCatching {
            StudyScheduleStore.shared(context).snapshot(LanguageScope.active(context)).upcoming(0L).map { it.at }
        }.getOrDefault(emptyList())
        val all = cards + study
        if (all.isEmpty()) return
        val hasDueNow = all.any { it <= now }
        val nextDue = all.filter { it > now }.minOrNull()
        val fireAt = fireDate(now, nextDue, hasDueNow) ?: return
        if (fireAt <= now) return
        val countAtFire = all.count { it <= fireAt }
        if (countAtFire == 0) return
        am.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, fireAt, pending(context, countAtFire))
    }

    /** When to fire, clamped into the waking-hours window. */
    fun fireDate(now: Long, nextDue: Long?, hasDueNow: Boolean): Long? {
        val cal = Calendar.getInstance()
        val candidate: Long = when {
            // They can see today's queue right now — remind tomorrow morning.
            hasDueNow -> {
                cal.timeInMillis = now
                cal.add(Calendar.DAY_OF_YEAR, 1)
                morning(cal)
            }
            nextDue != null -> nextDue
            else -> return null
        }
        // A due time inside the next 12 hours came from an explicit choice in
        // a bin tray — every ladder interval is a day or longer, and box 0 is
        // "due now" (handled above). Fire exactly then rather than parking a
        // short snooze until 9am.
        if (candidate - now < 12 * 3600_000L) return candidate
        cal.timeInMillis = candidate
        val hour = cal.get(Calendar.HOUR_OF_DAY)
        if (hour < DAY_START_HOUR) return morning(cal)
        if (hour >= DAY_END_HOUR) { cal.add(Calendar.DAY_OF_YEAR, 1); return morning(cal) }
        return candidate
    }

    private fun morning(cal: Calendar): Long {
        cal.set(Calendar.HOUR_OF_DAY, DAY_START_HOUR)
        cal.set(Calendar.MINUTE, 0); cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    private fun pending(context: Context, count: Int): PendingIntent =
        PendingIntent.getBroadcast(context, REQUEST_CODE,
            Intent(context, DrillDueReceiver::class.java).putExtra("count", count),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)

    fun notifyDue(context: Context, count: Int) {
        ReviewQueue.ensureChannel(context)
        val open = PendingIntent.getActivity(context, REQUEST_CODE,
            Intent(context, MainActivity::class.java)
                .putExtra(ReviewQueue.OPEN_REVIEW_EXTRA, true)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        // Counts everything waiting at fire time, not just the snooze that
        // triggered it — so the body says "waiting", not "you asked for this".
        val body = if (count == 1) context.getString(R.string.s_1_word_phrase_or_line_is_waiting)
        else context.getString(R.string.lld_words_phrases_and_lines_are_waiting, count)
        val n = NotificationCompat.Builder(context, ReviewQueue.CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_menu_recent_history)
            .setContentTitle(context.getString(R.string.ready_to_review))
            .setContentText(body).setAutoCancel(true).setContentIntent(open).build()
        (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager)
            .notify(NOTIFICATION_ID, n)
    }
}

class DrillDueReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        DrillReminder.notifyDue(context, intent.getIntExtra("count", 1).coerceAtLeast(1))
    }
}
